reframe/listener.go

```diff
@@ -0,0 +1,564 @@
+	// beginning of the list. To identify CIDs to expire, ReframeListener would walk the list tail to head.
+	// TODO: offload cid chunks to disk to save RAM
+	chunkByContextId       map[string]*cidsChunk
+	chunkByCid             map[cid.Cid]*cidsChunk
+	nodeByCid              map[cid.Cid]*cidNode
+	firstNode              *cidNode
+	lastNode               *cidNode
+	nonceGen               func() []byte
+	lastSeenProviderInfo   *peer.AddrInfo
+	configuredProviderInfo *peer.AddrInfo
+}
+
+// NewReframeListener creates a reframe listener with default random nonce generator
+func NewReframeListener(ctx context.Context, e provider.Interface, cidTtl time.Duration, chunkSize int, providerId string, addresses []string, ds datastore.Datastore) (*ReframeListener, error) {
+	return NewReframeListenerWithNonceGen(ctx, e, cidTtl, chunkSize, providerId, addresses, ds, func() []byte {
+		nonce := make([]byte, 8)
+		rand.Read(nonce)
+		return nonce
+	})
+}
+
+// NewReframeListenerWithNonceGen creates a reframe listener and initialises its state from the provided datastore. Nonce generator is used for calculating
+// context ids for new chunks.
+func NewReframeListenerWithNonceGen(ctx context.Context, e provider.Interface, cidTtl time.Duration, chunkSize int, providerId string, addresses []string, ds datastore.Datastore, nonceGen func() []byte) (*ReframeListener, error) {
+
+	indexProvider := &ReframeListener{e: e,
+		cidTtl:                 cidTtl,
+		chunkSize:              chunkSize,
+		currentChunk:           &cidsChunk{Cids: make(map[cid.Cid]bool)},
+		nodeByCid:              make(map[cid.Cid]*cidNode),
+		chunkByContextId:       make(map[string]*cidsChunk),
+		chunkByCid:             make(map[cid.Cid]*cidsChunk),
+		ds:                     ds,
+		nonceGen:               nonceGen,
+		lastSeenProviderInfo:   &peer.AddrInfo{},
+		configuredProviderInfo: nil,
+	}
+
+	e.RegisterMultihashLister(func(ctx context.Context, p peer.ID, contextID []byte) (provider.MultihashIterator, error) {
+		chunk := indexProvider.chunkByContextId[contextIDToStr(contextID)]
+		if chunk == nil {
+			log.Error("MultihasLister couldn't find chunk for contextID %s", contextIDToStr(contextID))
+			return nil, errors.New("MultihasLister couldn't find chunk for contextID")
+		}
+		var mhs []multihash.Multihash
+		for c := range chunk.Cids {
+			mhs = append(mhs, c.Hash())
+		}
+		return provider.SliceMultihashIterator(mhs), nil
+	})
+
+	indexProvider.initialiseFromTheDatastore(ctx)
+
+	maddrs, err := stringsToAddrs(addresses)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(providerId) != 0 {
+		p, err := peer.Decode(providerId)
+		if err != nil {
+			return nil, err
+		}
+		indexProvider.configuredProviderInfo = &peer.AddrInfo{
+			ID:    p,
+			Addrs: maddrs,
+		}
+	}
+
+	return indexProvider, nil
+}
+
+// initialiseFromTheDatastore initialises in-memory data structures on first start
+func (d *ReframeListener) initialiseFromTheDatastore(ctx context.Context) error {
```

Comment:
this has grown to a pretty complicated class

I wonder if the wrapper around / interaction with the datastore can be factored out to a second file / class to make the reframe provide logic clearer?

Reply:
Great point - moved all datastore related operations to a separate file `ds_wrapper`